homeassistant/components/picnic/services.py

```diff
@@ -0,0 +1,109 @@
+"""Services for the Picnic integration."""
+from __future__ import annotations
+
+from python_picnic_api import PicnicAPI
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import device_registry
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ATTR_AMOUNT,
+    ATTR_DEVICE_ID,
+    ATTR_PRODUCT_ID,
+    ATTR_PRODUCT_IDENTIFIERS,
+    ATTR_PRODUCT_NAME,
+    CONF_API,
+    DOMAIN,
+    SERVICE_ADD_PRODUCT_TO_CART,
+)
+
+
+class PicnicServiceException(Exception):
+    """Exception for Picnic services."""
+
+
+async def async_register_services(hass: HomeAssistant) -> None:
+    """Register services for the Picnic integration, if not registered yet."""
+
+    if hass.services.has_service(DOMAIN, SERVICE_ADD_PRODUCT_TO_CART):
+        return
+
+    async def async_add_product_service(call: ServiceCall):
+        # Get the picnic API client and call the handler with api client and hass
+        api_client = await get_api_client(hass, call.data.get("device_id"))
+        await handle_add_product(hass, api_client, call)
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_ADD_PRODUCT_TO_CART,
+        async_add_product_service,
+        schema=vol.Schema(
+            {
+                vol.Optional(ATTR_DEVICE_ID): cv.string,
+                vol.Exclusive(
+                    ATTR_PRODUCT_ID, ATTR_PRODUCT_IDENTIFIERS
+                ): cv.positive_int,
+                vol.Exclusive(ATTR_PRODUCT_NAME, ATTR_PRODUCT_IDENTIFIERS): cv.string,
+                vol.Optional(ATTR_AMOUNT): vol.All(vol.Coerce(int), vol.Range(min=1)),
+            }
+        ),
+    )
+
+
+async def get_api_client(
+    hass: HomeAssistant, device_id: str | None = None
+) -> PicnicAPI:
+    """Get the right Picnic API client based on the device id, else get the default one."""
+    if device_id is None:
+        default_config_id = list(hass.data[DOMAIN].keys())[0]
+        return hass.data[DOMAIN][default_config_id][CONF_API]
+
+    # Get device from registry
+    registry = device_registry.async_get(hass)
+    if not (device := registry.async_get(device_id)):
+        raise PicnicServiceException(f"Device with id {device_id} not found!")
+
+    # Get Picnic API client for the config entry id
+    try:
+        config_entry_id = next(iter(device.config_entries))
+        return hass.data[DOMAIN][config_entry_id][CONF_API]
+    except StopIteration as error:
+        raise PicnicServiceException(
+            f"Device with id {device_id} not found!"
+        ) from error
```

Comment:
Can be shortened without throwing an exception:

```suggestion
    if config_entry_id := next(iter(device.config_entries), None):
        return hass.data[DOMAIN][config_entry_id][CONF_API]
    raise PicnicServiceException(f"Device with id {device_id} not found!")
```